12-offer-encoding.md

```diff
@@ -0,0 +1,1056 @@
+3. The merchant makes a payment to the user indicated by the invoice.
+
+## Payment Proofs and Payer Proofs
+
+Note that the normal lightning "proof of payment" can only demonstrate that an
+invoice was paid (by showing the preimage of the `payment_hash`), not who paid
+it.  The merchant can claim an invoice was paid, and once revealed, anyone can
+claim they paid the invoice, too.[1]
+
+Providing a key in *invoice_request* allows a user to prove that they were the one
+to request the invoice.  In addition, the Merkle construction of the BOLT 12
+invoice signature allows the user to selectively reveal fields of the invoice
+in case of dispute.
+
+# Encoding
+
+Each of the forms documented here are in
+[TLV](01-messaging.md#type-length-value-format) format.
+
+The supported ASCII encoding is the human-readable prefix, followed by a
+`1`, followed by a bech32-style data string of the TLVs in order,
+optionally interspersed with `+` (for indicating additional data is to
+come).
+
+## Requirements
+
+Readers of a bolt12 string:
+- if it encounters a `+` followed zero or more whitespace characters between 
+  two bech32 characters:
+  - MUST remove the `+` and whitespace.
+
+## Rationale
+
+The use of bech32 is arbitrary, but already exists in the bitcoin
+world.  We currently omit the six-character trailing checksum: QR
+codes have their own checksums, bech32 doesn't protect against many
+length differences, and bech32m is not yet widely supported.
+
+The use of `+` (which is ignored) allows use over limited
+text fields like Twitter:
+
+```
+lno1xxxxxxxx+
+
+yyyyyyyyyyyy+
+
+zzzzz
+```
+
+See [format-string-test.json](bolt12/format-string-test.json).
+
+## Signature Calculation
+
+All signatures are created as per
+[BIP-340](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki),
+and tagged as recommended there.  Thus we define H(`tag`,`msg`) as
+SHA256(SHA256(`tag`) || SHA256(`tag`) || `msg`), and SIG(`tag`,`msg`,`key`)
+as the signature of H(`tag`,`msg`) using `key`.
+
+Each form is signed using one or more TLV signature elements; TLV
+types 240 through 1000 are considered signature elements.  For these
+the tag is "lightning" || `messagename` || `fieldname`, and `msg` is the
+Merkle-root; "lightning" is the literal 9-byte ASCII string,
+`messagename` is the name of the TLV stream being signed (i.e. "offer", "invoice_request" or "invoice") and the `fieldname` is the TLV field containing the
+signature (e.g. "signature" or "payer_signature").
+
+The formulation of the Merkle tree is similar to that proposed in
+[BIP-taproot], with each TLV leaf paired with a nonce leaf to avoid
+revealing adjacent nodes in proofs (assuming there is a non-revealed TLV
+which has enough entropy).
+
+The Merkle tree's leaves are, in TLV-ascending order for each tlv:
+1. The H(`LnLeaf`,tlv).
+2. The H(`LnAll`||all-tlvs,tlv) where "all-tlvs" consists of all non-signature TLV entries appended in ascending order.
```

Comment:
>> I think it should be called LnNonce instead of LnAll, it makes it clearer what this is for.
>
> Sure, but it's kind of arbitrary, and since I'd have to handle both old and new in my existing deployments I prefer to leave it as is.

Names are always arbitrary but "LnAll" is a terrible name, and you know it because in the example below you call it "nonce" instead (`L1nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV1)`). Changing the name now may break your existing deployments but that's fair game (DO NOT GET AN OFFER TATTOO UNTIL SPEC IS FULLY RATIFIED!), if not why bother with discussing the spec at all if nothing can change?

Reply:
Looking back, I realize that my message was a bit harsh, I'm sorry. The "LnAll" issue is very minor and if you really think that "LnAll" is better than "LnNonce" I'm fine with it. My concern is that we should aim to reach the best possible version of offer while the feature is still **experimental** and even for such a minor issue I think that breaking existing experimental deployments of offers is completely irrelevant.